File: app/src/main/java/com/nulchat/crypto/DoubleRatchetSession.java
package com.nulchat.crypto;

import java.security.GeneralSecurityException;
import java.util.HashMap;
import java.util.Map;

/**
 * See the original Kotlin version's doc comment (kept here for continuity)
 * for the full list of simplifications vs. the real Double Ratchet spec:
 * no X3DH, in-memory-only skipped-key cache, sessions not yet persisted
 * across app restarts. One addition versus the Kotlin version: the message
 * header (sender ratchet public key, previous chain length, message number)
 * is now bound into the AEAD's associated data, so tampering with those
 * fields in transit breaks authentication instead of silently corrupting
 * the ratchet state.
 */
public final class DoubleRatchetSession {

    private byte[] rootKey;
    private byte[] dhSelfSecret;
    private byte[] dhSelfPublic;
    private byte[] dhRemotePublic; // null until the first message is seen
    private byte[] sendChainKey;   // null until this side has ratcheted for sending
    private byte[] recvChainKey;   // null until this side has ratcheted for receiving
    private int sendMessageNumber = 0;
    private int recvMessageNumber = 0;
    private int previousSendChainLength = 0;

    private final Map<String, byte[]> skippedMessageKeys = new HashMap<>();

    private DoubleRatchetSession(byte[] rootKey, byte[] dhSelfSecret, byte[] dhSelfPublic) {
        this.rootKey = rootKey;
        this.dhSelfSecret = dhSelfSecret;
        this.dhSelfPublic = dhSelfPublic;
    }

    public static final class Envelope {
        public final byte[] senderRatchetPublicKey;
        public final int previousChainLength;
        public final int messageNumber;
        public final byte[] ciphertext; // Tink AEAD output — nonce is embedded

        public Envelope(byte[] senderRatchetPublicKey, int previousChainLength, int messageNumber, byte[] ciphertext) {
            this.senderRatchetPublicKey = senderRatchetPublicKey;
            this.previousChainLength = previousChainLength;
            this.messageNumber = messageNumber;
            this.ciphertext = ciphertext;
        }
    }

    public static DoubleRatchetSession initAsInitiator(byte[] sharedSecret, byte[] theirRatchetPublicKey)
            throws GeneralSecurityException {
        byte[] selfSecret = CryptoPrimitives.generateX25519SecretKey();
        byte[] selfPublic = CryptoPrimitives.x25519PublicFromSecret(selfSecret);
        DoubleRatchetSession session = new DoubleRatchetSession(sharedSecret, selfSecret, selfPublic);
        session.dhRatchetStep(theirRatchetPublicKey);
        return session;
    }

    public static DoubleRatchetSession initAsResponder(byte[] sharedSecret, byte[] ourRatchetSecretKey, byte[] ourRatchetPublicKey) {
        return new DoubleRatchetSession(sharedSecret, ourRatchetSecretKey, ourRatchetPublicKey);
    }

    public Envelope encrypt(byte[] plaintext) throws GeneralSecurityException {
        if (sendChainKey == null) throw new IllegalStateException("Session not initialized for sending");
        byte[][] step = symmetricRatchetStep(sendChainKey);
        byte[] messageKey = step[0];
        sendChainKey = step[1];

        int messageNumber = sendMessageNumber;
        byte[] aad = buildAad(dhSelfPublic, previousSendChainLength, messageNumber);
        byte[] ciphertext = CryptoPrimitives.encrypt(messageKey, plaintext, aad);

        Envelope envelope = new Envelope(dhSelfPublic, previousSendChainLength, messageNumber, ciphertext);
        sendMessageNumber += 1;
        return envelope;
    }

    public byte[] decrypt(Envelope envelope) throws GeneralSecurityException {
        String remoteKeyHex = toHex(envelope.senderRatchetPublicKey);
        String skippedKeyId = remoteKeyHex + ":" + envelope.messageNumber;

        byte[] skippedKey = skippedMessageKeys.remove(skippedKeyId);
        if (skippedKey != null) {
            byte[] aad = buildAad(envelope.senderRatchetPublicKey, envelope.previousChainLength, envelope.messageNumber);
            return CryptoPrimitives.decrypt(skippedKey, envelope.ciphertext, aad);
        }

        if (dhRemotePublic == null || !java.util.Arrays.equals(envelope.senderRatchetPublicKey, dhRemotePublic)) {
            skipRecvMessageKeys(envelope.previousChainLength);
            dhRatchetStep(envelope.senderRatchetPublicKey);
        }

        skipRecvMessageKeys(envelope.messageNumber);

        if (recvChainKey == null) throw new IllegalStateException("Session not initialized for receiving");
        byte[][] step = symmetricRatchetStep(recvChainKey);
        byte[] messageKey = step[0];
        recvChainKey = step[1];
        recvMessageNumber += 1;

        byte[] aad = buildAad(envelope.senderRatchetPublicKey, envelope.previousChainLength, envelope.messageNumber);
        return CryptoPrimitives.decrypt(messageKey, envelope.ciphertext, aad);
    }

    private void skipRecvMessageKeys(int until) throws GeneralSecurityException {
        if (recvChainKey == null || dhRemotePublic == null) return;
        byte[] current = recvChainKey;
        int n = recvMessageNumber;
        String remoteKeyHex = toHex(dhRemotePublic);
        while (n < until) {
            byte[][] step = symmetricRatchetStep(current);
            skippedMessageKeys.put(remoteKeyHex + ":" + n, step[0]);
            current = step[1];
            n += 1;
        }
        recvChainKey = current;
        recvMessageNumber = n;
    }

    private void dhRatchetStep(byte[] newRemotePublic) throws GeneralSecurityException {
        previousSendChainLength = sendMessageNumber;
        sendMessageNumber = 0;
        recvMessageNumber = 0;
        dhRemotePublic = newRemotePublic;

        byte[] recvShared = CryptoPrimitives.diffieHellman(dhSelfSecret, newRemotePublic);
        byte[][] recvStep = rootKdfStep(rootKey, recvShared);
        rootKey = recvStep[0];
        recvChainKey = recvStep[1];

        dhSelfSecret = CryptoPrimitives.generateX25519SecretKey();
        dhSelfPublic = CryptoPrimitives.x25519PublicFromSecret(dhSelfSecret);
        byte[] sendShared = CryptoPrimitives.diffieHellman(dhSelfSecret, newRemotePublic);
        byte[][] sendStep = rootKdfStep(rootKey, sendShared);
        rootKey = sendStep[0];
        sendChainKey = sendStep[1];
    }

    private byte[][] rootKdfStep(byte[] currentRootKey, byte[] dhOutput) throws GeneralSecurityException {
        byte[] nextRootKey = CryptoPrimitives.deriveKey(dhOutput, currentRootKey, "nulroot", 32);
        byte[] chainKey = CryptoPrimitives.deriveKey(dhOutput, currentRootKey, "nulchain", 32);
        return new byte[][] { nextRootKey, chainKey };
    }

    private byte[][] symmetricRatchetStep(byte[] chainKey) throws GeneralSecurityException {
        byte[] messageKey = CryptoPrimitives.deriveKey(chainKey, new byte[32], "nulmsg", 32);
        byte[] nextChainKey = CryptoPrimitives.deriveKey(chainKey, new byte[32], "nulchn", 32);
        return new byte[][] { messageKey, nextChainKey };
    }

    private static byte[] buildAad(byte[] senderRatchetPublicKey, int previousChainLength, int messageNumber) {
        byte[] aad = new byte[senderRatchetPublicKey.length + 8];
        System.arraycopy(senderRatchetPublicKey, 0, aad, 0, senderRatchetPublicKey.length);
        int offset = senderRatchetPublicKey.length;
        aad[offset] = (byte) (previousChainLength >> 24);
        aad[offset + 1] = (byte) (previousChainLength >> 16);
        aad[offset + 2] = (byte) (previousChainLength >> 8);
        aad[offset + 3] = (byte) previousChainLength;
        aad[offset + 4] = (byte) (messageNumber >> 24);
        aad[offset + 5] = (byte) (messageNumber >> 16);
        aad[offset + 6] = (byte) (messageNumber >> 8);
        aad[offset + 7] = (byte) messageNumber;
        return aad;
    }

    private static String toHex(byte[] data) {
        StringBuilder sb = new StringBuilder(data.length * 2);
        for (byte b : data) {
            sb.append(Character.forDigit((b >> 4) & 0xF, 16));
            sb.append(Character.forDigit(b & 0xF, 16));
        }
        return sb.toString();
    }
}

File: composeApp/src/commonMain/kotlin/com/nulchat/crypto/DoubleRatchetSession.kt
package com.nulchat.crypto

/**
 * A simplified Double Ratchet (Signal Protocol) session for one 1:1
 * conversation with a single peer.
 *
 * Simplifications vs. the full spec (worth revisiting post-MVP):
 * - No proper X3DH handshake yet. The very first shared secret is whatever
 *   [initAsInitiator]/[initAsResponder] are given — in this scaffold that's
 *   a DH between long-term identity-derived X25519 keys exchanged out of
 *   band (e.g. via the server invite link). A real X3DH would add signed
 *   prekeys + one-time prekeys for asynchronous, forward-secure first
 *   contact even while the peer is offline.
 * - Skipped/out-of-order message keys are kept only in memory
 *   ([skippedMessageKeys]), not persisted. If the app restarts before a
 *   delayed message arrives, that message becomes undecryptable. Persisting
 *   this map (encrypted) is a natural next step.
 * - This session object itself is not yet persisted to disk between app
 *   launches; see NulChatRepository/RatchetSession table for where that
 *   would plug in.
 *
 * Message numbering and the ratchet steps otherwise follow the standard
 * Double Ratchet: a DH ratchet advances the root key whenever a new remote
 * ratchet public key is seen, and a symmetric-key ratchet advances the
 * sending/receiving chain keys for every message within a chain.
 */
class DoubleRatchetSession private constructor(
    private var rootKey: ByteArray,
    private var dhSelfSecret: ByteArray,
    private var dhSelfPublic: ByteArray,
    private var dhRemotePublic: ByteArray?,
    private var sendChainKey: ByteArray?,
    private var recvChainKey: ByteArray?,
    private var sendMessageNumber: Int = 0,
    private var recvMessageNumber: Int = 0,
    private var previousSendChainLength: Int = 0
) {
    private val skippedMessageKeys = mutableMapOf<Pair<String, Int>, ByteArray>()

    data class Envelope(
        val senderRatchetPublicKey: ByteArray,
        val previousChainLength: Int,
        val messageNumber: Int,
        val nonce: ByteArray,
        val ciphertext: ByteArray
    )

    fun encrypt(plaintext: ByteArray): Envelope {
        val chainKey = sendChainKey ?: error("Session not initialized for sending")
        val (messageKey, nextChainKey) = symmetricRatchetStep(chainKey)
        sendChainKey = nextChainKey

        val nonce = CryptoPrimitives.randomNonce()
        val ciphertext = CryptoPrimitives.encrypt(messageKey, nonce, plaintext)
        val envelope = Envelope(
            senderRatchetPublicKey = dhSelfPublic,
            previousChainLength = previousSendChainLength,
            messageNumber = sendMessageNumber,
            nonce = nonce,
            ciphertext = ciphertext
        )
        sendMessageNumber += 1
        return envelope
    }

    fun decrypt(envelope: Envelope): ByteArray {
        val remoteKeyHex = envelope.senderRatchetPublicKey.toHex()

        skippedMessageKeys.remove(remoteKeyHex to envelope.messageNumber)?.let { key ->
            return CryptoPrimitives.decrypt(key, envelope.nonce, envelope.ciphertext)
        }

        if (dhRemotePublic == null || !envelope.senderRatchetPublicKey.contentEquals(dhRemotePublic)) {
            skipRecvMessageKeys(envelope.previousChainLength)
            dhRatchetStep(envelope.senderRatchetPublicKey)
        }

        skipRecvMessageKeys(envelope.messageNumber)

        val chainKey = recvChainKey ?: error("Session not initialized for receiving")
        val (messageKey, nextChainKey) = symmetricRatchetStep(chainKey)
        recvChainKey = nextChainKey
        recvMessageNumber += 1

        return CryptoPrimitives.decrypt(messageKey, envelope.nonce, envelope.ciphertext)
    }

    private fun skipRecvMessageKeys(until: Int) {
        val chainKey = recvChainKey ?: return
        var current = chainKey
        var n = recvMessageNumber
        val remoteKeyHex = (dhRemotePublic ?: return).toHex()
        while (n < until) {
            val (messageKey, nextChainKey) = symmetricRatchetStep(current)
            skippedMessageKeys[remoteKeyHex to n] = messageKey
            current = nextChainKey
            n += 1
        }
        recvChainKey = current
        recvMessageNumber = n
    }

    /** Advances the DH ratchet: new remote public key seen -> derive fresh root + chain keys. */
    private fun dhRatchetStep(newRemotePublic: ByteArray) {
        previousSendChainLength = sendMessageNumber
        sendMessageNumber = 0
        recvMessageNumber = 0
        dhRemotePublic = newRemotePublic

        // 1) Receiving chain: DH(ourCurrentSecret, theirNewPublic)
        val recvShared = CryptoPrimitives.diffieHellman(dhSelfSecret, newRemotePublic)
        val (newRootKeyAfterRecv, newRecvChain) = rootKdfStep(rootKey, recvShared)
        rootKey = newRootKeyAfterRecv
        recvChainKey = newRecvChain

        // 2) Generate a fresh sending ratchet keypair and derive the new sending chain
        dhSelfSecret = CryptoPrimitives.generateX25519SecretKey()
        dhSelfPublic = CryptoPrimitives.x25519PublicFromSecret(dhSelfSecret)
        val sendShared = CryptoPrimitives.diffieHellman(dhSelfSecret, newRemotePublic)
        val (newRootKeyAfterSend, newSendChain) = rootKdfStep(rootKey, sendShared)
        rootKey = newRootKeyAfterSend
        sendChainKey = newSendChain
    }

    private fun rootKdfStep(currentRootKey: ByteArray, dhOutput: ByteArray): Pair<ByteArray, ByteArray> {
        // Root KDF: mix the DH output into the root key to get (nextRootKey, chainKey).
        // Using the DH output itself as the KDF master key, keyed by the current root key
        // via subkey id/context, keeps this a single call into CryptoPrimitives.deriveKey.
        val nextRootKey = CryptoPrimitives.deriveKey(dhOutput, subkeyId = 1, context = "nulroot")
        val chainKey = CryptoPrimitives.deriveKey(dhOutput, subkeyId = 2, context = "nulchain")
        return nextRootKey to chainKey
    }

    private fun symmetricRatchetStep(chainKey: ByteArray): Pair<ByteArray, ByteArray> {
        val messageKey = CryptoPrimitives.deriveKey(chainKey, subkeyId = 1, context = "nulmsg_")
        val nextChainKey = CryptoPrimitives.deriveKey(chainKey, subkeyId = 2, context = "nulchn_")
        return messageKey to nextChainKey
    }

    companion object {
        /**
         * Call this on the side that sends the first message. [sharedSecret] must be
         * agreed out of band (see class doc — X3DH is the real-world way to do this).
         */
        fun initAsInitiator(sharedSecret: ByteArray, theirRatchetPublicKey: ByteArray): DoubleRatchetSession {
            val selfSecret = CryptoPrimitives.generateX25519SecretKey()
            val selfPublic = CryptoPrimitives.x25519PublicFromSecret(selfSecret)
            val session = DoubleRatchetSession(
                rootKey = sharedSecret,
                dhSelfSecret = selfSecret,
                dhSelfPublic = selfPublic,
                dhRemotePublic = null,
                sendChainKey = null,
                recvChainKey = null
            )
            session.dhRatchetStep(theirRatchetPublicKey)
            return session
        }

        /** Call this on the side that waits to receive the first message. */
        fun initAsResponder(
            sharedSecret: ByteArray,
            ourRatchetSecretKey: ByteArray,
            ourRatchetPublicKey: ByteArray
        ): DoubleRatchetSession {
            return DoubleRatchetSession(
                rootKey = sharedSecret,
                dhSelfSecret = ourRatchetSecretKey,
                dhSelfPublic = ourRatchetPublicKey,
                dhRemotePublic = null,
                sendChainKey = null,
                recvChainKey = null
            )
        }
    }
}

private fun ByteArray.toHex(): String {
    val hexChars = "0123456789abcdef"
    val sb = StringBuilder(size * 2)
    for (b in this) {
        val v = b.toInt() and 0xFF
        sb.append(hexChars[v shr 4])
        sb.append(hexChars[v and 0x0F])
    }
    return sb.toString()
}
